File: native/app/src/main/java/com/example/masterannouncer/service/Service.kt
package com.example.masterannouncer.service

import androidx.lifecycle.LiveData
import androidx.room.Update
import com.example.masterannouncer.domain.Announcement
import com.example.masterannouncer.domain.AnnouncementId
import com.example.masterannouncer.repository.Repository
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import okhttp3.WebSocket
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*

object Service {

    //Running from the emulator, start the server first.
    private const val URL = "http://10.0.2.2:3000/"
    lateinit var repository: Repository

    interface Service {
        @GET("/announcements")
        suspend fun getAnnouncements(): List<Announcement>

        @POST("/announcements")
        suspend fun postAnnouncement(@Body announcement: Announcement): Announcement

        @DELETE("/announcements")
        suspend fun deleteAnnouncement(@Query("id") id: Long)

        @PUT("/announcements")
        suspend fun updateAnnouncement(@Body announcement: Announcement)
    }

    private val interceptor: HttpLoggingInterceptor = HttpLoggingInterceptor().apply {
        this.level = HttpLoggingInterceptor.Level.BODY
    }

    private val client: OkHttpClient = OkHttpClient.Builder().apply {
        this.addInterceptor(interceptor)
    }.build()


    private var gson = GsonBuilder()
        .setLenient()
        .create()

    private val retrofit = Retrofit.Builder()
        .baseUrl(URL)
        .addConverterFactory(GsonConverterFactory.create(gson))
        .client(client)
        .build()

    val service: Service = retrofit.create(Service::class.java)
    

}

File: non-native/master_announcer/lib/service/service.dart


import 'package:master_announcer/domain/announcement.dart';
import 'package:master_announcer/repository/repository.dart';

class Service{

  static final Service _service=Service._internal();

  Repository repository;

  factory Service(){
    return _service;
  }

  Service._internal(){
    repository=Repository();
  }


  Future<int> addAnnouncement(Announcement a){
    return repository.addAnnouncement(a);
  }

  Future<int> addAnnouncementId(Announcement a){
    return repository.addAnnouncementId(a);
  }

  void deleteAnnouncement(int id){
    repository.deleteAnnouncement(id);
  }

  void editAnnouncement(Announcement a){
    repository.editAnnouncement(a);
  }


  Future<List<Announcement>> getAnnouncements() async {
    List<Announcement> a= await repository.getAnnouncements();
    return a;
  }

  void addToAdd(int id){
    repository.addToAdd(id);
  }

  void deleteAll(){
    repository.deleteAll();
  }

  Future<List<int>> getToAdd() {
    return repository.getToAdd();
}
}